compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp

```diff
@@ -0,0 +1,105 @@
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/SourceMgr.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/PDL/IR/PDL.h"
+#include "mlir/Dialect/PDLInterp/IR/PDLInterp.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Transform/IR/TransformDialect.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/FileUtilities.h"
+#include "mlir/Support/LLVM.h"
+
+static llvm::cl::opt<std::string> clTransformFileName(
+    "dispatch-transform-dialect-file-name",
+    llvm::cl::desc("mlir file containing a top-level module that specifies "
+                   "the transformations to apply to form dispatch regions."),
+    llvm::cl::init(""));
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+/// Pass declaration.
+struct DispatchWithTransformDialect
+    : public DispatcWithTransformDialectBase<DispatchWithTransformDialect> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    // clang-format off
+    registry.insert<IREE::Flow::FlowDialect,
+                    AffineDialect,
+                    arith::ArithmeticDialect,
+                    linalg::LinalgDialect,
+                    pdl::PDLDialect,
+                    pdl_interp::PDLInterpDialect,
+                    scf::SCFDialect,
+                    tensor::TensorDialect
+    >();
+    // clang-format on
+  }
+  DispatchWithTransformDialect() = default;
+  DispatchWithTransformDialect(const DispatchWithTransformDialect &pass) {}
+  void runOnOperation() override;
+
+ private:
+  Statistic numDispatches{this, "number of dispatches",
+                          "Number of Flow dispatches created"};
+};
+
+void DispatchWithTransformDialect::runOnOperation() {
+  Operation *topLevel = getOperation();
+  if (topLevel->getNumRegions() != 1 ||
+      !llvm::hasSingleElement(topLevel->getRegion(0))) {
+    topLevel->emitError() << "can only run '" << getArgument()
+                          << "' on single-region single-block operations";
+    return signalPassFailure();
+  }
+
+  if (clTransformFileName.empty()) {
+    topLevel->emitError() << "expected a transform file name";
+    return signalPassFailure();
+  }
+
+  // Parse clTransformFileName content into a ModuleOp.
+  std::string errorMessage;
+  auto memoryBuffer = mlir::openInputFile(clTransformFileName, &errorMessage);
+  if (!memoryBuffer) {
+    llvm::errs() << errorMessage << "\n";
+    return signalPassFailure();
+  }
+
+  // Tell sourceMgr about this buffer, the parser will pick it up.
+  llvm::SourceMgr sourceMgr;
+  sourceMgr.AddNewSourceBuffer(std::move(memoryBuffer), llvm::SMLoc());
+  OwningOpRef<ModuleOp> transformModule(
+      parseSourceFile<ModuleOp>(sourceMgr, &getContext()));
```

Comment:
Is this thread-safe? (this pass will be called concurrently)

We should at least memoize it - this isn't a cheap operation and when using the transform dialect reusing the transformModule is going to be important anywhere it's used.

Reply:
Good point!
Note that this is definitely not ready for prime time but only to allow connecting pieces for now.
Do you have a suggestion as to where that memoization should occur?
I can lift this into the initializer for now but this leads to a bad use of a shared_ptr due to various fun things around MLIR passes.
Please also see my batch reply/question for more discussion.